### PR TITLE
[Doc] better documention

### PR DESCRIPTION
**Pull request by archibate**
Related issue = #736 

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

### CODE REVIEW

File: docs/atomic.rst [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+This will be translated to (in GPU pseudo-code):
+::
+    read from ``result`` to register 1
+    read from ``x[i]`` to register 2
+    add the value of reg1 and reg2, save result to reg3
+    write back the value in register 3 to ``result``
+
+This code is executed by many parallel GPU threads for different ``i``.
+What if thread 233 is doing add, while thread 666 has already write back the ``result``?
```
Comment (k-ye) [inline docs/atomic.rst:21]: Let's be serious and use `1` and `2`

File: docs/atomic.rst [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+            result[None] = result[None] + x[i]
+
+This will be translated to (in GPU pseudo-code):
+::
+    read from ``result`` to register 1
+    read from ``x[i]`` to register 2
+    add the value of reg1 and reg2, save result to reg3
+    write back the value in register 3 to ``result``
+
+This code is executed by many parallel GPU threads for different ``i``.
```
Comment (k-ye) [inline docs/atomic.rst:20]: A high level comment - I have a feeling that most users are aware of what atomics are, and what problem atomics can solve. Maybe just mention the fact that Taichi supports atomic operations on both integers and floats?

File: docs/atomic.rst [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        for i in x:
+            ti.atomic_add(result[None], x[i])
+
+
+Functions
+---------
+
+.. function:: ti.atomic_add(x, y)
+
+    This is equivalent with ``x += y``.
```
Comment (k-ye) [inline docs/atomic.rst:52]: equivalent to

File: docs/atomic.rst [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    ::
+        x = 3
+        y = 4
+        z = ti.atomic_add(x, y)
+        # now x = 7, y = 4, z = 3
+
+
+.. function:: ti.atomic_sub(x, y)
+
+    This is equivalent with ``x -= y``.
```
Comment (k-ye) [inline docs/atomic.rst:68]: equivalent to

File: docs/atomic.rst [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+        z = ti.atomic_max(x, y)
+        # now x = 4, y = 4, z = 3
+
+
+.. function:: ti.atomic_min(x, y)
+
+
+.. function:: ti.atomic_or(x, y)
+
+    This is equivalent with ``x |= y``.
```
Comment (k-ye) [inline docs/atomic.rst:86]: equivalent to

File: docs/atomic.rst [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+
+.. function:: ti.atomic_and(x, y)
+
+    This is equivalent with ``x &= y``.
```
Comment (k-ye) [inline docs/atomic.rst:91]: equivalent to

File: docs/atomic.rst [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+
+
+.. function:: ti.atomic_xor(x, y)
+
+    This is equivalent with ``x ^= y``.
```
Comment (k-ye) [inline docs/atomic.rst:96]: equivalent to

File: docs/cpp_style.rst [hunk 8]
```diff
@@ -28,3 +28,7 @@ Don'ts
 - ``NULL``, use ``nullptr`` instead.
 - ``using namespace std;`` in global scope.
 - ``typedef``. Use ``using`` instead.
+
+Automatically format
```
Comment (k-ye) [inline docs/cpp_style.rst:32]: Automatic code formatting

File: docs/vector.rst [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+    :return: (Vector) the outer product of ``a`` and ``b``
+
+    e.g.:
+    ::
+        a = ti.Vector([1, 2, 3])
+        b = ti.Vector([4, 5, 6])
+        c = ti.outer_product(a, b) # [1*4, 2*5, 3*6]
+
+.. note::
+    This have not common with `ti.cross`. It simply multiplies each compoments. And thus ``a`` and ``b`` does not have to be 3D vectors.
```
Comment (k-ye) [inline docs/vector.rst:128]: This is not the same as

File: docs/vector.rst [hunk 10]
```diff
@@ -0,0 +1,141 @@
+.. _vector:
+
+Vector
+======
+
+
+``ti.Vector`` is actually a kind of ``ti.var``, but have multiple compoment variables within it.
```
Comment (k-ye) [inline docs/vector.rst]: a convenient wrapper over?

File: docs/vector.rst [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+
+    :parameter a: (Vector, 3D)
+    :parameter b: (Vector, 3D)
+    :return: (Vector, 3D) the cross product of ``a`` and ``b``
+
+    e.g.:
+    ::
+        a = ti.Vector([1, 2, 3])
+        b = ti.Vector([4, 5, 6])
+        c = ti.cross(a, b) # [2*6 - 5*3, 4*3 - 1*6, 1*5 - 4*2]
```
Comment (archibate) [inline docs/vector.rst:112]: Was this correct? We may want to tell them if we use left-handed or right-handed?

File: docs/vector.rst (hunk 10 above)
Comment (k-ye) [inline docs/vector.rst]: I think your next paragraph has explained the idea, so maybe this is duplicative?

File: docs/vector.rst [hunk 12]
```diff
@@ -0,0 +1,141 @@
+.. _vector:
+
+Vector
+======
+
+
+``ti.Vector`` is actually a kind of ``ti.var``, but have multiple compoment variables within it.
+A ``n``-D vector is composed of ``n`` variables. It's just a syntax sugar to abstract a linear algebra framework, no real difference in lower execution for now.
```
Comment (k-ye) [inline docs/vector.rst]: ... framework. There is ...

File: docs/atomic.rst [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+    def sum():
+        for i in x:
+            result[None] = result[None] + x[i]
+
+This will be translated into (GPU pseudo-code):
+::
+    read from ``result`` to register 1
+    read from ``x[i]`` to register 2
+    add the value of reg1 and reg2, save result to reg3
+    write back the value in register 3 to ``result``
```
Comment (archibate) [inline docs/atomic.rst:18]: I noticed that we haven't doced atomic operation... and I add this.

File: docs/utilities.rst [hunk 14]
```diff
@@ -3,6 +3,30 @@ Utilities
...
+
+.. code-block:: python
+
+    gui = ti.GUI('Title', (640, 480))
+    while not gui.is_pressed(ti.GUI.ESCAPE):
+        gui.set_image(img)
+        gui.show()
+
+
+Also checkout https://github.com/taichi-dev/taichi/blob/master/examples/keyboard.py for more advanced event processing.
```
Comment (archibate) [inline docs/utilities.rst:17]: TODO: we want a more detailed API desc like atomic and vector does.

File: docs/atomic.rst (hunk 3 above)
Comment (archibate) [reply docs/atomic.rst:52]: Sorry for my bad English!

File: docs/atomic.rst (hunk 2 above)
Comment (archibate) [reply docs/atomic.rst:20]: I thought we taichi is a repo for getting CG newcomers?
> ...on both integers and floats...
OK, let's also add a table for supported atomic operation of different backends.

**Comment by archibate** [comment]
> ...native speakers as reviewer...

Any native speaker in @taichi-dev? Maybe we can ask @yuanming-hu for help?

**Comment by yuanming-hu** [comment]
Let me take a pass first.

File: docs/vector.rst [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+
+    :parameter a: (Vector)
+    :parameter b: (Vector)
+    :return: (Vector) the outer product of ``a`` and ``b``
+
+    e.g.:
+    ::
+        a = ti.Vector([1, 2, 3])
+        b = ti.Vector([4, 5, 6])
+        c = ti.outer_product(a, b) # [1*4, 2*5, 3*6]
```
Comment (yuanming-hu) [inline docs/vector.rst:124]: `[1*4, 2*5, 3*6]` is the element-wise product, represented as `a * b` in Taichi. 

`ti.outer_product(a, b)` generates `[[1 * 4, 1 * 5, 1 * 6], [2 * 4, 2 * 5, 2 * 6], [3 * 4, 3 * 5, 3 * 6]]`. See https://en.wikipedia.org/wiki/Outer_product

File: docs/atomic.rst (hunk 2 above)
Comment (yuanming-hu) [reply docs/atomic.rst:20]: I think here we should refer readers to https://en.wikipedia.org/wiki/Fetch-and-add for an explanation of `atomic_add` and simply say `x[i] += 1` is atomic but `x[i] = x[i] + 1` is not. Explaining what atomic operations mean should probably not be part of taichi's documentation. (But feel free to write a blog post with these details! People will appreciate that :-)

File: docs/vector.rst (hunk 15 above)
Comment (archibate) [reply docs/vector.rst]: Oh, I understand what it means now, it's $a^T b$ for inner-product, and $a b^T$ for outer-product.

Comment (yuanming-hu) [reply docs/vector.rst]: Exactly!!

**Comment by yuanming-hu** [comment]
One more high-level thing: should we use in-place documentation?  @archibate @k-ye 
 https://pythonhosted.org/an_example_pypi_project/sphinx.html#function-definitions

This will allow us to write documentation together with code. 

I think we can have two parts: tutoriala and a more complete API reference. Tutorials are written by human and API reference generated automatically.

**Comment by archibate** [comment]
> One more high-level thing: should we use in-place documentation? @archibate @k-ye
> https://pythonhosted.org/an_example_pypi_project/sphinx.html#function-definitions
> 
> This will allow us to write documentation together with code.
> 
> I think we can have two parts: tutoriala and a more complete API reference. Tutorials are written by human and API reference generated automatically.

Sounds like a `doxygen` will do? But I still want to keep some human-written reference for things like `ti.Vector([x, y, ...])`, which could be hard if generated by machine(?)

**Comment by yuanming-hu** [comment]
> Sounds like a `doxygen` will do?

Right, we use `sphinx` here for the Python part.

 

> But I still want to keep some human-written reference for things like `ti.Vector([x, y, ...])`, which could be hard if generated by machine(?)

I agree - that's why I believe we should have both tutorials and API references.

**Comment by archibate** [comment]
Fine, ;et's keep some critical parts to be human-written, like the definition of tensor&vectors. And non-critical parts like the API of GUI, we can machine gen it.

File: docs/type.rst [hunk 16]
```diff
@@ -5,11 +5,47 @@ Supported types
...
+    | u16  |    OK     |   MISS    |   OK    |
+    +------+-----------+-----------+---------+
+    | u32  |    OK     |   MISS    |   OK    |
+    +------+-----------+-----------+---------+
+    | u64  |    OK     |   MISS    |  MISS   |
+    +------+-----------+-----------+---------+
+    | f32  |    OK     |    OK     |   OK    |
+    +------+-----------+-----------+---------+
+    | f64  |    OK     |   EXT     |  MISS   |
+    +------+-----------+-----------+---------+
```
Comment (archibate) [inline docs/type.rst:44]: Please verify if I told correct on your backends @yuanming-hu @k-ye. Thanks!

File: docs/atomic.rst [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+    | type | LLVM      | OpenGL    | Metal   |
+    +======+===========+===========+=========+
+    | i32  |    OK     |    OK     |   OK    |
+    +------+-----------+-----------+---------+
+    | f32  |    OK     |    OK     |   OK    |
+    +------+-----------+-----------+---------+
+    | i64  |    OK     |   EXT     |  MISS   |
+    +------+-----------+-----------+---------+
+    | f64  |    OK     |   EXT     |  MISS   |
+    +------+-----------+-----------+---------+
```
Comment (archibate) [inline docs/atomic.rst:38]: And here.

File: docs/hello.rst [hunk 18]
```diff
@@ -8,16 +8,10 @@ Depending on your hardware and OS, please execute one of the following commands:
...
+  # python 3.6+ required
 
-  # CPU only. No GPU/CUDA needed. (Linux, OS X and Windows)
   python3 -m pip install taichi-nightly
 
-  # With GPU (CUDA 10.0) support (Linux only)
-  python3 -m pip install taichi-nightly-cuda-10-0
-
-  # With GPU (CUDA 10.1) support (Linux only)
-  python3 -m pip install taichi-nightly-cuda-10-1
```
Comment (archibate) [inline docs/hello.rst:20]: I removed them since I heard #714.

File: docs/type.rst (hunk 16 above)
Comment (yuanming-hu) [reply docs/type.rst:44]: LLVM looks good! Thanks for adding the table. It's very helpful.

File: docs/atomic.rst (hunk 17 above)
Comment (yuanming-hu) [reply docs/atomic.rst:38]: LLVM looks good. Btw should we replace `LLVM` with `CPU/CUDA`? Most users probably don't know what LLVM is.

File: docs/vector.rst (hunk 11 above)
Comment (yuanming-hu) [reply docs/vector.rst:111]: Yes, this is correct. Let's make it clear that we are using our right hands, as most people do in this world.

File: docs/utilities.rst (hunk 14 above)
Comment (yuanming-hu) [reply docs/utilities.rst:17]: We can always do a simple tutorial plus a complete API reference. For `GUI` this also applies :-)

File: docs/hello.rst (hunk 18 above)
Comment (yuanming-hu) [reply docs/hello.rst:20]: I'll try to have v0.5.11 released with `taichi-nightly` covering all CUDA versions as soon as possible.

**Comment by archibate** [comment]
Btw, I would like to add a page describing what cards does `ti.cuda` support, i.e. Pascal cards, or 900+ cards?

**Comment by archibate** [comment]
Do we have `clang-9` support now?

**Comment by yuanming-hu** [comment]
Before I do my pass, do you mind doing me a favor and fix these Sphinx warnings?

```
/home/yuanming/repos/taichi/docs/atomic.rst:8: WARNING: Unexpected indentation.
/home/yuanming/repos/taichi/docs/atomic.rst:15: WARNING: Unexpected indentation.
/home/yuanming/repos/taichi/docs/atomic.rst:27: WARNING: Unexpected indentation.
/home/yuanming/repos/taichi/docs/atomic.rst:34: WARNING: Unexpected indentation.
/home/yuanming/repos/taichi/docs/atomic.rst:41: WARNING: Unexpected indentation.
/home/yuanming/repos/taichi/docs/atomic.rst:60: WARNING: Unexpected indentation.
/home/yuanming/repos/taichi/docs/atomic.rst:75: WARNING: Unexpected indentation.
/home/yuanming/repos/taichi/docs/contributor_guide.rst:80: WARNING: Inline interpreted text or phrase reference start-string without end-string.
/home/yuanming/repos/taichi/docs/contributor_guide.rst:80: WARNING: Inline interpreted text or phrase reference start-string without end-string.
/home/yuanming/repos/taichi/docs/hello.rst:96: WARNING: Title underline too short.

(Sparse) Tensors
-------
/home/yuanming/repos/taichi/docs/hello.rst:96: WARNING: Title underline too short.

(Sparse) Tensors
-------
/home/yuanming/repos/taichi/docs/scalar_tensor.rst:17: WARNING: Unexpected indentation.
/home/yuanming/repos/taichi/docs/scalar_tensor.rst:21: WARNING: Unexpected indentation.
/home/yuanming/repos/taichi/docs/scalar_tensor.rst:25: WARNING: Unexpected indentation.
/home/yuanming/repos/taichi/docs/vector.rst:24: WARNING: Unexpected indentation.
/home/yuanming/repos/taichi/docs/vector.rst:42: WARNING: Unexpected indentation.
/home/yuanming/repos/taichi/docs/vector.rst:56: WARNING: Unexpected indentation.
/home/yuanming/repos/taichi/docs/vector.rst:60: WARNING: Unexpected indentation.
/home/yuanming/repos/taichi/docs/vector.rst:76: WARNING: Unexpected indentation.
/home/yuanming/repos/taichi/docs/vector.rst:81: WARNING: Unexpected indentation.
/home/yuanming/repos/taichi/docs/vector.rst:96: WARNING: Unexpected indentation.
/home/yuanming/repos/taichi/docs/vector.rst:109: WARNING: Unexpected indentation.
/home/yuanming/repos/taichi/docs/vector.rst:122: WARNING: Unexpected indentation.
/home/yuanming/repos/taichi/docs/vector.rst:138: WARNING: Unexpected indentation.
```

**Comment by archibate** [comment]
> Before I do my pass, do you mind doing me a favor and fix these Sphinx warnings? No rush on this. Thanks!

No problem! Fixed all WARNINGs except for contributor_guideline.rst, how to fix that? Seems related to the link.

**Comment by yuanming-hu** [comment]
Thanks! I'm taking a pass. Adding a space after `_` seems to fix the remaining warning.

File: docs/dev_install.rst [hunk 19]
```diff
@@ -19,39 +22,62 @@ Note that on Linux/OS X, ``clang`` is the only supported compiler for compiling
...
+    cd llvm-8.0.1.src
     mkdir build
     cd build
     cmake .. -DLLVM_ENABLE_RTTI:BOOL=ON -DBUILD_SHARED_LIBS:BOOL=OFF -DCMAKE_BUILD_TYPE=Release -DLLVM_TARGETS_TO_BUILD="X86;NVPTX" -DLLVM_ENABLE_ASSERTIONS=ON
     # If you are building on NVIDIA Jetson TX2, use -DLLVM_TARGETS_TO_BUILD="ARM;NVPTX"
     make -j 8
     sudo make install
 
-- Clone the taichi repo, and then
+Setting up Taichi for develop
```
Comment (yuanming-hu) [inline docs/dev_install.rst:53]: develop->development

File: docs/dev_install.rst [hunk 20]
```diff
@@ -19,39 +22,62 @@ Note that on Linux/OS X, ``clang`` is the only supported compiler for compiling
...
+---------------------------------------------
+
+- Clone the taichi repo, and build:
 
   .. code-block:: bash
 
+    git clone https://github.com/taichi-dev/taichi --depth=1 --branch=master
+    gsm update --init --recursive --depth=1
```
Comment (yuanming-hu) [inline docs/dev_install.rst:61]: gsm -> git submodule

File: docs/dev_install.rst [hunk 21]
```diff
@@ -19,39 +22,62 @@ Note that on Linux/OS X, ``clang`` is the only supported compiler for compiling
...
     cd taichi
     mkdir build
     cd build
     cmake ..
-    # if you are building with CUDA, say, 10.0, then please use "cmake .. -DCUDA_VERSION=10.0 -DTI_WITH_CUDA:BOOL=True"
+    # if you are building with CUDA 10.0, use this:
```
Comment (yuanming-hu) [inline docs/dev_install.rst:66]: this -> the line below

File: docs/dev_install.rst [hunk 22]
```diff
@@ -19,39 +22,62 @@ Note that on Linux/OS X, ``clang`` is the only supported compiler for compiling
...
+    # cmake .. -DCUDA_VERSION=10.0 -DTI_WITH_CUDA:BOOL=True
     make -j 8
 
-- Add the following to your ``~/.bashrc`` (or ``~/.zshrc`` if you use ``zsh``)
+- Add the following codes to your ``~/.bashrc``:
```
Comment (yuanming-hu) [inline docs/dev_install.rst:70]: codes -> script. Note that `codes` is almost never used.

File: docs/global_settings.rst [hunk 23]
```diff
@@ -3,4 +3,8 @@ Global Settings
 
 - Restart the Taichi runtime system (clear memory, destroy all variables and kernels): ``ti.reset()``
 - Eliminate verbose outputs: ``ti.get_runtime().set_verbose(False)``
-- To specify which GPU to use: ``export CUDA_VISIBLE_DEVICES=0``
+- Not to trigger GDB when crashes: ``ti.set_gdb_trigger(False)``
```
Comment (yuanming-hu) [inline docs/global_settings.rst:6]: Not to -> To not

File: docs/hello.rst [hunk 24]
```diff
@@ -65,16 +63,35 @@ You can also reuse the package management system, Python IDEs, and existing Pyth
 Portability
 -----------------
 
-Taichi supports both CPUs and NVIDIA GPUs.
+Taichi code can run on CPUs or GPUs. Depends on your platform:
 
 .. code-block:: python
 
-  # Run on GPU
+  # Run on NVIDIA GPU, require CUDA installed
```
Comment (yuanming-hu) [inline docs/hello.rst:70]: require CUDA installed -> CUDA required

File: docs/hello.rst [hunk 25]
```diff
@@ -65,16 +63,35 @@ You can also reuse the package management system, Python IDEs, and existing Pyth
 Portability
 -----------------
 
-Taichi supports both CPUs and NVIDIA GPUs.
+Taichi code can run on CPUs or GPUs. Depends on your platform:
```
Comment (yuanming-hu) [inline docs/hello.rst:66]: Depends on your platform -> Initialize Taichi according to your hardware platform

File: docs/hello.rst [hunk 26]
```diff
@@ -65,16 +63,35 @@ You can also reuse the package management system, Python IDEs, and existing Pyth
...
 
 .. code-block:: python
 
-  # Run on GPU
+  # Run on NVIDIA GPU, require CUDA installed
   ti.init(arch=ti.cuda)
+  # Run on GPU, with OpenGL backend
```
Comment (yuanming-hu) [inline docs/hello.rst:72]: OpenGL backend -> the OpenGL backend

File: docs/hello.rst [hunk 27]
```diff
@@ -65,16 +63,35 @@ You can also reuse the package management system, Python IDEs, and existing Pyth
...
+  ti.init(arch=ti.opengl)
+  # Run on GPU, with Metal backend, if you're on OS X
```
Comment (yuanming-hu) [inline docs/hello.rst:74]: Metal backend -> the Metal backend
you're -> you are (In general, avoid `I'll`, `you're`, `can't` in formal writing)

File: docs/hello.rst [hunk 28]
```diff
@@ -65,16 +63,35 @@ You can also reuse the package management system, Python IDEs, and existing Pyth
...
+  ti.init(arch=ti.metal)
   # Run on CPU (default)
   ti.init(arch=ti.x64)
 
-If the machine does not have CUDA support, Taichi will fall back to CPUs instead.
+.. note::
+    Supported archs on different platforms:
```
Comment (yuanming-hu) [inline docs/hello.rst:80]: archs -> backends

File: docs/hello.rst [hunk 29]
```diff
@@ -65,16 +63,35 @@ You can also reuse the package management system, Python IDEs, and existing Pyth
...
+
+    +----------+------+------+--------+-------+
+    | platform | CPU  | CUDA | OpenGL | Metal |
+    +==========+======+======+========+=======+
+    | Windows  | OK   | WIP  | WIP    | MISS  |
```
Comment (yuanming-hu) [inline docs/hello.rst:85]: Windows CUDA is `OK`

File: docs/hello.rst [hunk 30]
```diff
@@ -65,16 +63,35 @@ You can also reuse the package management system, Python IDEs, and existing Pyth
...
+    +----------+------+------+--------+-------+
+    | Linux    | OK   | OK   | OK     | MISS  |
+    +----------+------+------+--------+-------+
+    | Mac OS X | OK   | MISS | MISS   | OK    |
+    +----------+------+------+--------+-------+
+
+    (OK=supported, WIP=work in progress, MISS=not supported)
```
Comment (yuanming-hu) [inline docs/hello.rst:92]: Use `N/A` instead of `MISS` since we won't support them in the future.

File: docs/hello.rst [hunk 31]
```diff
@@ -171,9 +188,23 @@ In the fractal code above, ``for i, j in pixels`` loops over all the pixel coord
           for i in x:
             ...
 
-.. warning::
+.. note::
+    ``break`` is not supported in **outermost (parallelized)** loops:
 
-    Struct-for's must be at the outer-most scope of kernels.
```
Comment (yuanming-hu) [inline docs/hello.rst:176]: Shall we keep this line?

File: docs/linalg.rst [hunk 32]
```diff
@@ -1,13 +1,12 @@
...
-===============================================
-
 Matrices
----------------------------------------
+========
+
 - ``ti.Matrix`` is for small matrices (e.g. `3x3`) only. If you have `64x64` matrices, you should consider using a 2D tensor of scalars.
 - ``ti.Vector`` is the same as ``ti.Matrix``, except that it has only one column.
 - Differentiate element-wise product ``*`` and matrix product ``@``.
+- ``ti.Vector(n, dt=ti.f32)`` or ``ti.Matrix(n, m, dt=ti.f32)``.
```
Comment (yuanming-hu) [inline docs/linalg.rst:9]: please append `to create tensors of vectors/matrices`.

File: docs/hello.rst (hunk 31 above)
Comment (archibate) [reply docs/hello.rst:176]: I thought this is (and can be) supported?
```
import taichi as ti

x = ti.var(ti.i32, 3)
y = ti.var(ti.i32, 3)

@ti.kernel
def func():
    for i in x:
        for j in y:
            x[i] *= y[j]

func()
```

Comment (yuanming-hu) [reply docs/hello.rst:176]: No, this is not supported.

**Comment by archibate** [comment]
Travis can't download openssl? Excuse me?

**Comment by yuanming-hu** [comment]
Finalized the remaining two files. Thanks for contributing to the documentation!

It's hard to make perfect, but hopefully, readers will catch and fix our errors later.

**Pull request merged**